ui/store/actions.js

```diff
@@ -1038,19 +1039,59 @@ export function updateMetamaskState(newState) {
   return (dispatch, getState) => {
     const { metamask: currentState } = getState();
 
-    const { currentLocale, selectedAddress } = currentState;
+    const { currentLocale, selectedAddress, provider } = currentState;
     const {
       currentLocale: newLocale,
       selectedAddress: newSelectedAddress,
+      provider: newProvider,
     } = newState;
 
     if (currentLocale && newLocale && currentLocale !== newLocale) {
       dispatch(updateCurrentLocale(newLocale));
     }
+
     if (selectedAddress !== newSelectedAddress) {
       dispatch({ type: actionConstants.SELECTED_ADDRESS_CHANGED });
     }
 
+    // Ensuring that the chainId is not undefined for both providers ensures we
+    // can rely upon the logic of our getMetaMaskAccounts selector instead of
+    // replicating its behavior here in a way more tolerant of missing
+    // properties
+    if (provider?.chainId !== undefined && newProvider?.chainId !== undefined) {
```

Comment:
Is it possible for `chainId` to be undefined? :thinking:

Reply:
Which is good, because if it was possible, I think this condition would result in missed events for account changed and selected account changed.

Reply:
If it ever is the other selectors and state tree accessing will fail with a visible error to user. I ran into a number of these cases during testing which could be remedied by making test case state mocks more robust but leaves the code more vulnerable to edge cases where the chainId is not defined.

If you're confident in it always being present I can remove these checks and increase the robustness of the test cases.

Reply:
I am fairly confident - and if it's every unset, that's a pretty severe bug. Ensuring it's always set was one of the goals of Erik's cross-product chain ID validation work. There are security reasons to ensure it's always set.

The only remaining state that I'm aware of that _sometimes_ has a chain ID but not always is hold transaction history.

Reply:
resolved in: https://github.com/MetaMask/metamask-extension/pull/10965/commits/7a9c778af47765c25a49a41336d80c2a86e2dfd9 -- weird but no test failures so maybe my concern was from a point in the past of this behemoth.